`|=` does not work as expected on TOMLDocument

I was expecting `|=` operator to work the same as `.update` method, because that's how Python dicts work, but `|=` on a TOMLDocument seems to have no effect:

```python
from tomlkit import TOMLDocument

d1 = TOMLDocument()
d1['a'] = {'b': 'c'}
d2 = TOMLDocument()
d2['a'] = {'b': 'Z'}
d1['a'] |= d2['a']  # expected to be the same as d1['a'].update(d2['a']), but it's not

assert d1 == {'a': {'b': 'Z'}}, d1  # AssertionError: {'a': {'b': 'c'}}
```

Same code using a Python dict:
```python
d1 = {'a': {'b': 'c'}}
d2 = {'a': {'b': 'Z'}}

d1['a'] |= d2['a']

assert d1 == {'a': {'b': 'Z'}}  # no error
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,10 @@
 
 ## [Unreleased]
 
+### Fixed
+
+- Support `|` and `|=` operator for tables, and support `+` and `+=` operator for arrays. ([#331](https://github.com/sdispater/tomlkit/issues/331))
+
 ## [0.12.3] - 2023-11-15
 
 ### Fixed

--- a/tomlkit/_types.py
+++ b/tomlkit/_types.py
@@ -43,8 +43,26 @@ else:
     class _CustomList(MutableSequence, list):
         """Adds MutableSequence mixin while pretending to be a builtin list"""
 
+        def __add__(self, other):
+            new_list = self.copy()
+            new_list.extend(other)
+            return new_list
+
+        def __iadd__(self, other):
+            self.extend(other)
+            return self
+
     class _CustomDict(MutableMapping, dict):
         """Adds MutableMapping mixin while pretending to be a builtin dict"""
+
+        def __or__(self, other):
+            new_dict = self.copy()
+            new_dict.update(other)
+            return new_dict
+
+        def __ior__(self, other):
+            self.update(other)
+            return self
 
     class _CustomInt(Integral, int):
         """Adds Integral mixin while pretending to be a builtin int"""
@@ -54,7 +72,7 @@ else:
 
 
 def wrap_method(
-    original_method: Callable[Concatenate[WT, P], Any]
+    original_method: Callable[Concatenate[WT, P], Any],
 ) -> Callable[Concatenate[WT, P], Any]:
     def wrapper(self: WT, *args: P.args, **kwargs: P.kwargs) -> Any:
         result = original_method(self, *args, **kwargs)

